packages/flutter_tools/lib/src/runner/flutter_command.dart

```diff
@@ -695,8 +696,15 @@ abstract class FlutterCommand extends Command<void> {
   Future<void> validateCommand() async {
     if (_requiresPubspecYaml && !PackageMap.isUsingCustomPackagesPath) {
       // Don't expect a pubspec.yaml file if the user passed in an explicit .packages file path.
-      if (!globals.fs.isFileSync('pubspec.yaml')) {
-        throw ToolExit(userMessages.flutterNoPubspec);
+
+      // If there is no pubspec in the current directory, look in the parent
+      // until one can be found.
+      while (!globals.fs.isFileSync('pubspec.yaml')) {
+        final Directory nextCurrent = globals.fs.currentDirectory.parent;
+        if (nextCurrent == null || nextCurrent.path == globals.fs.currentDirectory.path) {
+          throw ToolExit(userMessages.flutterNoPubspec);
+        }
+        globals.fs.currentDirectory = nextCurrent;
```

Comment:
If this loop succeeds, it would probably help if we printed a status saying what directory we are using as the cwd.

Reply:
Only if we are _not_ using the cwd we started with

Reply:
We should not ever print non-actionable messages, see https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#only-log-actionable-messages-to-the-console

Reply:
The action would be to change your cwd if something has gone wrong, right?

Reply:
Imagine, for example, the following directory structures:

```
src/
  pubspec.yaml           // put here last week when I didn't trust git to not lose it on me
  flutter_app_1/
    pubspec.yaml
  java_app_1/
    // no pubspec.yaml
```

One day I do `flutter run` in `java_app_1` and to my surprise weird things happen. I would have found it really helpful to know that it was because the tool found a stray pubspec.yaml that I left laying around on my FS.

Alternatively, imagine I did `flutter run -t ../lib/main.dart` from the `android` directory. And imagine that, for some crazy reason, I have a `../../lib/main.dart` present in my filesystem.  The log message would have made it clearer which main.dart was actually getting used.

Reply:
It would change the cwd of the dart process, but nothing else is modified.

Reply:
Won't the dart process at least potentially be resolving relative paths, such as the `-t` option to flutter run?

Reply:
Sure, but those might be wrong anyway. At any rate, we're still strictly better than we were before:

1. flutter run  (in lib/): failed
2. flutter run --target=main.dart (in lib/): failed

After:

1. flutter run (in lib/): passed
2. flutter run --target=main.dart (in lib/): failed


If this leads to confusion, we can always tweak the error messaging

Reply:
What's the output of `flutter run --target=main.dart` in `lib/` with this patch?

Reply:
Target file "main.dart" not found.

Reply:
Okay, I'll back out my comment on non-actionable message :) . Like you suggested, we status log to the new CWD if it is updated, then the error message above is more interpretable